src/core/lib/event_engine/grpc_ares_wrapper.h

```diff
@@ -0,0 +1,340 @@
+// Copyright 2023 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#ifndef GRPC_SRC_CORE_LIB_EVENT_ENGINE_GRPC_ARES_WRAPPER_H
+#define GRPC_SRC_CORE_LIB_EVENT_ENGINE_GRPC_ARES_WRAPPER_H
+
+#include <grpc/support/port_platform.h>
+
+#if GRPC_ARES == 1
+
+#include <stddef.h>
+
+#include <algorithm>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include <ares.h>
+
+#include "absl/base/thread_annotations.h"
+#include "absl/functional/any_invocable.h"
+#include "absl/status/status.h"
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+#include "absl/types/optional.h"
+
+#include <grpc/event_engine/event_engine.h>
+#include <grpc/support/log.h>
+
+#include "src/core/lib/debug/trace.h"
+#include "src/core/lib/event_engine/grpc_polled_fd.h"
+#include "src/core/lib/gprpp/orphanable.h"
+#include "src/core/lib/gprpp/sync.h"
+
+namespace grpc_event_engine {
+namespace experimental {
+
+extern grpc_core::TraceFlag grpc_trace_ares_wrapper;
+
+#define GRPC_ARES_WRAPPER_TRACE_LOG(format, ...)                              \
+  do {                                                                        \
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_trace_ares_wrapper)) {                   \
+      gpr_log(GPR_INFO, "(EventEngine c-ares wrapper) " format, __VA_ARGS__); \
+    }                                                                         \
+  } while (0)
+
+// Base class for the c-ares based DNS lookup functionality. This class is
+// inherited by the concrete leaf classes below which each represents a
+// different type of lookup (i.e. A/AAAA, SRV, TXT). One request object
+// represents one lookup request for a specific name. GrpcAresRequest
+// encapsulates the common logics to initialize and destroy the c-ares channel
+// (one channel for one request). Sets the name servers configuration for the
```

Comment:
This is an interesting point. I think it is true that c-ares API allows multiple requests on a single c-ares channel (it needs to be synchronized so multiple requests will share the same lock [0]). It might even save memory in doing so, e.g. per-channel linked-list|hash table and name servers list etc. can be shared.

However, there might be some hurdles in this approach:
1. The biggest one seems to be cancellation. [ares_cancel](https://c-ares.org/ares_cancel.html) will cancel all the queries associated with a channel and there seems to be no alternative (see [1] also). This is in conflict with the EventEngine API which allows to cancel each individual request.
2. Another sort of related issue is when we get a list of sockets/fds associated with a channel to wait on from c-ares, we don't necessarily know which one corresponding to which request (see [ares_getsock](https://c-ares.org/ares_getsock.html)). Besides the obvious issue with cancellation, this also has issue in timeout situation. Since we have separate timeout per-request, if one request timed out, we need to shutdown the poller handle and cancel the request. But we don't necessarily know which FdNodes should we shutdown for this request nor we can cancel the request per previous point.

[0] https://github.com/c-ares/c-ares/issues/208
[1] https://github.com/c-ares/c-ares/issues/135#issuecomment-392425642

Reply:
> What if we moved both cancellation and timeouts to the DNSResolver object, rather than doing them on the individual requests?

@markdroth You mean something like `DNSResolver::CancelAll()`? That sounds probably right from a client channel resolver perspective, but less-right from a generic resolver API perspective. If users want independent requests, they'd have to make a new DNSResolver for each request. It would be an awkward API, but usable. The bigger issue is that cancellation gets pretty weird if some subset of requests are not in a cancellable state, but others are - some client callbacks will be run but not others, and the caller would need to know what to expect based on whatever `CancelAll` returns.

> Don't take this optimization and live with the overhead of separate c-ares channels per request.

I think I agree with @apolcyn here, and would hope that an extra couple of `fd`s consumed while resolving TXT records would be tolerable.

> And we could probably live with expressing timeouts at the DNSResolver level as well.

This may be a moot point, but having a single timeout for multiple requests, all started at different times, could have a few interpretations. Does each request time out independently, or would the timeout apply as soon as the first request is sent? Alternatively, the creation of a DNSResolver could mean configuring a batch of requests, then starting them all at once. But again, I think this harms the API to serve a specific internal use case.

> I believe that the entire reason that we designed the EE API to put the DNS methods into a DNSResolver object was to share this kind of state.

If we don't share state, then we _could_ move the DNS lookup methods into the Engine API itself. But I still think having a separate object is nice for organization and design purposes, single-responsibility principle in particular.

Reply:
> Do we actually have any such use-cases [for a specific DNS cancellation usage]? I'm not aware of any.

No users specifically for DNS externally yet that I'm aware of, but I expect so eventually. We have external EventEngine users and authors already. And _we ourselves_ may eventually be users who want fine-grained cancellation, because it seems like the most natural usage of a generic resolver API. See below.

> Alternatively, maybe the semantic for the DNS callbacks should be that the callbacks are always run, even upon cancellation?

That would confuse users. I'd really prefer to not have two different cancellation semantics in the EventEngine API. If it simplified things immensely, the tradeoff might be worth it.

> This API is actually designed specifically for our use-cases, so I don't consider [tuning the API for an internal use case] a problem.

I disagree here. Tuning a public API for our current internal use cases (especially when it makes the API more awkward for any other use case) would be a mistake for everyone. This public API may very well outlive the specifics of what gRPC is doing internally with DNS. And we know the EventEngine is being used outside gRPC already, with more plans to create custom engines. I don't want to create an awkward future by optimizing for the wrong things today.

> If we actually have the ability to configure the DNS servers for each query, why aren't we doing it that way?

It's a good point, and we have that ability today with an Ares implementation. If you aren't using ares, specifying the authority on every request might not make sense. Having a different timeout for every request could make a lot of sense, though - say a lookup implementation returns DEADLINE_EXCEEDED, I can imagine wanting to retry a lookup with a longer timeout. The design here was generally that you create a DNS Resolver for a given authority, and with a given set of overall resolver settings ... and if you need to resolve against a different authority or with incompatible settings, you get a different resolver.  But If the usefulness of a separate `DNSResolver` object is in question, then it could be a moot point as well.

Another thing to consider is creating a separate, optional `GetBatchedDNSResolver` method with a no-op default. This could provide a mechanism for channel sharing optimizations without harming the generality of the API, and probably allow a lot of code reuse internally. I think the addition is a bit kludgy already, but if we need the optimization, we could.

Reply:
Chatting with @yijiem, the more we work through potential semantics for `GetBatchedDNSResolver`, the worse the idea seems. I retract my support for the concept.